### PR TITLE
Unify the computation of distances in Cartesian and spherical coordinates

### LINKED ISSUE #69: Unify computation of distances in spherical and Cartesian coordinates
**Description of the desired feature**

To prevent repeated code, would be nice to have single functions for computing distances between two points in both spherical and Cartesian coordinates. They must be Numba jitted functions to speed things up.

For the spherical coordinates case, the points are given in `longitude`, `latitude` and `radius` coordinates, but several trigonometric functions must be applied in order to ultimately get the distance between these two points. Some forward models precompute these quantities in order to save time. So, would be nice to have a private function that takes all these precomputed quantities as arguments and returns the distance, and a public function that just takes the points' coordinates, compute this quantities and use this private function to get the distance between them.

Moreover, some forward models just needs the distance square, so we could make another private function that just returns the distance square, i.e. doesn't apply the square root.

For example:
```python
@jit(nopython=True)
def distance_spherical(point_a, point_b):
    # Get coordinates of the two points
    longitude, latitude, radius = point_a[:]
    longitude_p, latitude_p, radius_p = point_b[:]
    # Convert angles to radians
    longitude, latitude = np.radians(longitude), np.radians(latitude)
    longitude_p, latitude_p = np.radians(longitude_p), np.radians(latitude_p)
    # Compute trigonometric quantities
    cosphi_p = np.cos(latitude_p)
    sinphi_p = np.sin(latitude_p)
    cosphi = np.cos(latitude)
    sinphi = np.sin(latitude)
    return _distance_spherical(
        longitude, cosphi, sinphi, radius, longitude_p, cosphi_p, sinphi_p, radius_p
    )


@jit(nopython=True)
def _distance_spherical(
    longitude, cosphi, sinphi, radius, longitude_p, cosphi_p, sinphi_p, radius_p
):
    return np.sqrt(
        _distance_sq_spherical(
            longitude, cosphi, sinphi, radius, longitude_p, cosphi_p, sinphi_p, radius_p
        )
    )


@jit(nopython=True)
def _distance_sq_spherical(
    longitude, cosphi, sinphi, radius, longitude_p, cosphi_p, sinphi_p, radius_p
):
    coslambda = np.cos(longitude_p - longitude)
    cospsi = sinphi_p * sinphi + cosphi_p * cosphi * coslambda
    return (radius - radius_p) ** 2 + 2 * radius * radius_p * (1 - cospsi)
```

**Are you willing to help implement and maintain this feature?** Yes

### PR DESCRIPTION
**Pull request by santisoler**
Add functions to compute distances in both Cartesian and spherical coordinates on
forward/utils.py. Make point mass and tesseroid forward models to use these functions.

Fixes #69 

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If adding new functionality, add an example to the docstring, gallery, and/or tutorials.

### CODE REVIEW

**Comment by santisoler** [comment]
@leouieda I'm thinking we could merge `distance_spherical()` and `distance_cartesian()` into a single `distance()` function that takes `coordinate_system` as an argument, as `point_mass_gravity()` does. What do you think?

**Comment by leouieda** [comment]
:+1: yes! API consistency :fireworks:

**Comment by santisoler** [comment]
@leouieda I merged the computation of distances into a single public function. I've also added a `_check_coordinate_system()` to unify the verification of valid coordinate system. It's currently used on `point_mass_gravity()` and `distance()`.

I had to make little changes on the private jitted functions because they are needed on the forward modelling functions.

Let me know if you see something that could be improved.

File: doc/api/index.rst [hunk 1]
```diff
@@ -24,6 +24,7 @@ Forward modelling
 
     point_mass_gravity
     tesseroid_gravity
+    distance
```
Comment (leouieda) [inline doc/api/index.rst]: I think this probably shouldn't be in the public AP if it's a utility meant for use in the forward modeling modules. We can always make it public later but removing the function is harder.

File: harmonica/forward/point_mass.py [hunk 2]
```diff
@@ -303,8 +287,8 @@ def kernel_g_r(
...
     Kernel function for radial component of gravity gradient in spherical coordinates
     """
-    coslambda = np.cos(longitude_p - longitude)
-    cospsi = sinphi_p * sinphi + cosphi_p * cosphi * coslambda
-    distance_sq = (radius - radius_p) ** 2 + 2 * radius * radius_p * (1 - cospsi)
+    distance_sq, cospsi, _ = _distance_sq_spherical_core(
+        longitude, cosphi, sinphi, radius, longitude_p, cosphi_p, sinphi_p, radius_p
+    )
     delta_z = radius_p * cospsi - radius
     return delta_z / distance_sq ** (3 / 2)
```
Comment (leouieda) [inline harmonica/forward/point_mass.py]: I learned a while ago that sometimes this is actually a lot slower than `np.sqrt(distance)**3`. So we don't really need the distance to be squared.

**Comment by santisoler** [comment]
> Hi Santi, I've been going through the code and it seems like the `distance` function isn't really used anywhere. The actual functions used are the jit functions. So these should not have `_`. The `_` indicates functions that aren't meant to be used outside of their module. 

Hi  @leouieda. Thanks for the review. I tend to get confused about *private* and *public* functions and if all public functions should be in the API. Thanks for making it clearer.

> And these are used across `forward`. Here is my proposal:
> 
>     1. Remove the `distance` function from the API. It isn't used right now and we can always add it later if necessary. No need to delete it but it's best to wait for a use case before making it public.

No problem!

>     2. Make the jit function "public" but not in the API.

I agree with this.

>     3. Put the `srqt` in the distances. Squaring something is a cheap operation and pretty much all functions do the sqrt at some point.

You are right, working the squared values will eventually create some mistakes.

File: harmonica/forward/point_mass.py (hunk 2 above)
Comment (santisoler) [reply harmonica/forward/point_mass.py]: :+1: Didn't know that. I thought that applying a single power to a value would be faster than applying two different ones in series. Although I have no references to hold this statement, just intuition.

**Comment by leouieda** [comment]
:+1: good to merge on my part. Looks great!

**Pull request merged**